### ISSUE TITLE
Unsoundness: Patterns in function parameters are not checked for union access

### ISSUE THREAD

**Issue by ChayimFriedman2** [rust-lang/rust#130528]
I tried this code:

```rust
union U {
    a: &'static i32,
    b: usize,
}

fn foo(U { a }: U) {
    dbg!(*a);
}

fn main() {
    foo(U { b: 0 });
}
```
[Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=84ba6b82e516b4b204d2ecd1bd846d65).

I expected to see this happen: rustc rejecting this code. We are transmuting arbitrary types without unsafe, this is clearly unsound.

Instead, this happened: rustc accepted this code, and it SIGSEGVs at runtime.

**Comment by Urgau** [comment]
Seems to have regressed between 1.76 and 1.77.

In 1.76 (and below) it gave:
```
error[E0133]: access to union field is unsafe and requires unsafe function or block
 --> <source>:6:12
  |
6 | fn foo(U { a }: U) {
  |            ^ access to union field
  |
  = note: the field may not be properly initialized: using uninitialized data will cause undefined behavior
```

**Comment by compiler-errors** [comment]
I'll take a look at this. I think I may have found out what's going on.

@rustbot claim

**Comment by apiraino** [comment]
WG-prioritization assigning priority ([Zulip discussion](https://rust-lang.zulipchat.com/#narrow/stream/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/.23130528.20Unsoundness.3A.20Patterns.20in.20function.20parameters.20are.20no.E2.80.A6)).

@rustbot label -I-prioritize +P-critical

**Comment by compiler-errors** [comment]
Reopening for beta nomination.

**Comment by wesleywiser** [comment]
Fix backported via #130531
